pkg/sql/catalog/table_elements.go

```diff
@@ -411,27 +411,85 @@ type Constraint interface {
 	// GetName returns the name of this constraint update mutation.
 	GetName() string
 
-	// NotNullColumnID returns the underlying not-null column ID, if there is one.
-	NotNullColumnID() descpb.ColumnID
-
 	// AsCheck returns the underlying check constraint, if there is one.
-	AsCheck() *descpb.TableDescriptor_CheckConstraint
+	AsCheck() CheckConstraint
 
 	// AsForeignKey returns the underlying foreign key constraint, if there is
 	// one.
-	AsForeignKey() *descpb.ForeignKeyConstraint
+	AsForeignKey() ForeignKeyConstraint
 
 	// AsUniqueWithoutIndex returns the underlying unique without index
 	// constraint, if there is one.
-	AsUniqueWithoutIndex() *descpb.UniqueWithoutIndexConstraint
+	AsUniqueWithoutIndex() UniqueWithoutIndexConstraint
 
-	// AsPrimaryKey returns the index descriptor backing the PRIMARY KEY
-	// constraint, if there is one.
-	AsPrimaryKey() Index
+	// AsUnique returns the index descriptor backing the UNIQUE or the
+	// PRIMARY KEY constraint, if there is one.
+	AsUnique() UniqueWithIndexConstraint
+}
+
+// UniqueWithIndexConstraint is an interface around a unique constraint
+// which backed by an index.
+type UniqueWithIndexConstraint interface {
+	Constraint
+	Index
+}
+
+// WithoutIndexConstraint is the supertype of all constraint subtypes which are
+// not backed by an index.
+type WithoutIndexConstraint interface {
+	Constraint
+}
+
+// CheckConstraint is an interface around a check constraint.
+type CheckConstraint interface {
+	WithoutIndexConstraint
+
+	// CheckDesc returns the underlying descriptor protobuf.
+	CheckDesc() *descpb.TableDescriptor_CheckConstraint
+
+	// Expr returns the check expression as a string.
+	Expr() string
+
+	// NumReferencedColumns returns the number of columns referenced in the check
+	// expression.
+	NumReferencedColumns() int
+
+	// GetReferencedColumnID returns the ID of the column referenced in the check
+	// expression at ordinal `columnOrdinal`.
+	GetReferencedColumnID(columnOrdinal int) descpb.ColumnID
```

Comment:
I see this pattern of having `NumReferencedColumns` and `GetReferencedColumnID` methods in both Check and FK constraint interface. Do you expect users to use these two methods together to retrieve all referenced columns? 

Why don't we add a method, say called `GetColumnIDs()` in `catalog.Constraint` interface? It will return
    - check: columns in the check expression
    - FK: columns in the original table (say the FK is from A --> B, then that method returns columns in A)
    - UniqueWithoutIndex: column(s) that are unique 
    - PK: the primary index's key columns
    - UNIQUE: this unique index's key columns

Reply:
We don't like to return slices because the ownership of the slice is not clear and it's easy to mess things up terribly by `append`-ing something. Most of the time we just want to loop through the slice contents or we want to check set properties, hence a long time ago we settled on this `NumColID/GetColID/CollectColIDs` pattern.